### PR TITLE
WT-13121 Create lock hierarchy page

### PR DESCRIPTION
**Pull request by ershov**

### CODE REVIEW

File: src/docs/arch-locks.dox [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+</table>
+
+<table class=lock-hierarchy>
+@row{WRITE \c log_remove_lock, &rarr;, READ \c debug_log_retention_lock}
+</table>
+
+NOTE: Some functions expect \c log_slot_lock to be acquired, then they temporarily release it and re-acquire it.
+
+@section arch-locks-checkpoint-schema Checkpoint and Schema locks
+
```
Comment (quchenhao) [inline src/docs/arch-locks.dox:61]: No description from this section. You intend to do it later?

Comment (ershov) [reply src/docs/arch-locks.dox:61]: I actually thought of removing the locks descriptions at all. The locks are well documented in the source, it's easy to find. The purpose of this page is to document the locking order.

Comment (ershov) [reply src/docs/arch-locks.dox:38]: Removed descriptions.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/docs/arch-locks.dox [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+@row{\c reconfig_lock, &rarr;, \c chunkcache->hashtable[*].bucket_lock, &rarr;, READ \c chunkcache->pinned_objects.array_lock}
+</table>
+
+@section arch-locks-examples Examples
+
+@subsection arch-locks-examples-eviction0 Eviction (no queue lock)
+
+@image html locks-evict-example-simple.svg width=900
+<!-- callgraph -f '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t __evict_clear_walk -t __evict_walk_tree -continue -tagrex '/(__wt_spin_[^u]\w++)\({ARG},\s++({ARG})/index($2, "evict") != -1 && index($2, "queue") == -1 ? "$1($2)" : undef' -tag "WT_WITH_PASS_LOCK(evict_pass_lock)=WT_WITH_PASS_LOCK" -x /_curhs_cache -x /_verbose_dump_cache -x /_cache_destroy -x /_cache_create -d 11 -dir td -dot | dot -Tsvg /dev/stdin > src/docs/images/locks-evict-example-simple.svg -->
+
```
Comment (marcbutler) [inline src/docs/arch-locks.dox:90]: Just confirming: this comment documents how the graph was created? Is callgraph in the code base?

File: src/docs/arch-index.dox [hunk 3]
```diff
@@ -165,6 +165,7 @@ of the documentation.
 
 - @subpage arch-toc-int-wt-dev
     - @ref arch-concurrency
+    - @ref arch-locks
```
[the remark this answers is not in the document]
Comment (ershov) [reply src/docs/arch-index.dox:168]: Good suggestion, thanks!

File: src/docs/arch-locks.dox (hunk 2 above)
Comment (ershov) [reply src/docs/arch-locks.dox:90]: It's not in the code base but this comment will help to re-build it in future.

File: src/docs/arch-index.dox (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcbutler) [reply src/docs/arch-index.dox:168]: Sorry I don't understand. You responded in the positive but the name unchanged.

[the remark this answers is not in the document]
Comment (ershov) [reply src/docs/arch-index.dox]: Ah sorry I forgot to push the change. Done now.

**Review by marcbutler — APPROVED** [review]
I don't have time to validate the analysis, but as it is mechanical it feels safe.

LGTM

File: src/docs/arch-locking-hierarchy.dox [hunk 4]
```diff
@@ -0,0 +1,105 @@
+/*! @arch_page arch-locking-hierarchy Locks hierarchy
+
+This page describes the current nested locks hierarchy of WiredTiger.
```
Comment (sueloverso) [inline src/docs/arch-locking-hierarchy.dox:3]: Minor, both here and line 1 I think "Locking hierarchy" flows better than "Locks hierarchy", and it matches the name of the file.

File: dist/s_docs [hunk 5]
```diff
@@ -225,6 +225,7 @@ EOF
 
     # Copy any needed javascript
     cp ../src/docs/js/* ../docs/
+    cp ../src/docs/*.html ../docs/
```
Comment (sueloverso) [inline dist/s_docs:228]: There have never been any pre-existing HTML files in the doc directory before. Why is this necessary?

File: src/docs/arch-locking-hierarchy.dox [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+<table class=lock-hierarchy>
+@row{\c reconfig_lock, &rarr;, \c chunkcache->hashtable[*].bucket_lock, &rarr;, READ \c chunkcache->pinned_objects.array_lock}
+</table>
+
+@section arch-locking-hierarchy-examples Examples
+
+@subsection arch-locking-hierarchy-examples-eviction0 Eviction (no queue lock)
+
+@image html locks-evict-example-simple.svg width=900
+<!-- callgraph -f '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t __evict_clear_walk -t __evict_walk_tree -continue -tagrex '/(__wt_spin_[^u]\w++)\({ARG},\s++({ARG})/index($2, "evict") != -1 && index($2, "queue") == -1 ? "$1($2)" : undef' -tag "WT_WITH_PASS_LOCK(evict_pass_lock)=WT_WITH_PASS_LOCK" -x /_curhs_cache -x /_verbose_dump_cache -x /_cache_destroy -x /_cache_create -d 11 -dir td -dot | dot -Tsvg /dev/stdin > src/docs/images/locks-evict-example-simple.svg -->
```
Comment (sueloverso) [inline src/docs/arch-locking-hierarchy.dox:89]: It feels like all of these commands could be fragile with any naming changes or code development and evolution. Can we just generate the image? Does the inclusion of these help the reader?

File: src/docs/images/locks-evict-example-simple.svg
Comment (sueloverso) [inline src/docs/images/locks-evict-example-simple.svg:1]: Everything else in the images directory is a PNG. Is it necessary to introduce another image format? Could these be PNGs?

File: src/docs/locks-eviction.html
Comment (sueloverso) [inline src/docs/locks-eviction.html:1]: There currently are no HTML files in the `src/docs` only `.dox` files. Is it necessary to introduce these?

**Review by sueloverso — COMMENTED** [review]
The content is mostly okay. I have a few comments about the infrastructure.

File: src/docs/arch-locking-hierarchy.dox (hunk 4 above)
Comment (ershov) [reply src/docs/arch-locking-hierarchy.dox]: Fixed

File: src/docs/arch-locking-hierarchy.dox (hunk 6 above)
Comment (ershov) [reply src/docs/arch-locking-hierarchy.dox]: Changed the commands to not include the `__`/`__wt`/`__wti` prefix.

The image is generated and included directly. These commands are hints how to re-generate them in future if anything changes.

File: src/docs/images/locks-evict-example-simple.svg
Comment (ershov) [reply src/docs/images/locks-evict-example-simple.svg:1]: There is a significant difference between PNG and SVG. SVG is the best fit for diagrams and vector graphics. SVG can infinitely scale up and down without any quality issues. SVG has been a standard for decades and is supported by all known browsers.
Are you concerned about unnecessarily introducing new entities or there are any other issues with it?

File: dist/s_docs (hunk 5 above)
Comment (ershov) [reply dist/s_docs:228]: Because the call graphs are rather big, these HTML pages also include a minimal script to pan and zoom it (similar to maps). I thought, it's a good visual illustration of the subject.

File: src/docs/locks-eviction.html
Comment (ershov) [reply src/docs/locks-eviction.html:1]: See my comment in s_docs

File: dist/s_docs (hunk 5 above)
Comment (sueloverso) [reply dist/s_docs:228]: Thanks for the explanation.

**Review by sueloverso — APPROVED** [review]
LGTM Thanks for the replies.

**Review by ddanderson — APPROVED** [review]
@ershov I took a quick look and generated the doc to try it out.  Looks very nice to me.  I made some minor comments.

File: src/docs/arch-eviction.dox [hunk 7]
```diff
@@ -86,29 +86,8 @@ cache contains at least this many bytes of updates.
...
 than its trigger size counterpart.
 
-@section eviction_locking Locking
-
-Internally eviction uses the following locks:
-
-- \c cache->evict_pass_lock - Guarantees that only one eviction server is running at most.
-- \c cache->evict_walk_lock - Locked when eviction walks a tree.
-- \c cache->evict_queue_lock - Access to cache->evict_queues.
-- \c queue->evict_lock - Access to a specific eviction queue.
```
[the remark this answers is not in the document]
Comment (ershov) [reply src/docs/arch-eviction.dox:96]: I intentionally removed descriptions of locks since they are just copied from the source. The purpose of the page is to document locking hierarchy which is sometimes hard to get. Locks descriptions are easy to find where they are declared and I didn't want to overload the page with unnecessary information.

If you believe that this kind of table would be useful, I can make a script that would automatically generate one from sources. This method will also guarantee that the table is always up to date.

File: src/docs/arch-locking-hierarchy.dox [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+<table class=lock-hierarchy>
+@row{\c reconfig_lock, &rarr;, \c chunkcache->hashtable[*].bucket_lock, &rarr;, READ \c chunkcache->pinned_objects.array_lock}
+</table>
+
+@section arch-locking-hierarchy-examples Examples
+
+@subsection arch-locking-hierarchy-examples-eviction0 Eviction (no queue lock)
+
+@image html locks-evict-example-simple.svg width=900
+<!-- callgraph -f '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t /_evict_clear_walk -t /_evict_walk_tree -continue -tagrex '/(__wt_spin_[^u]\w++)\({ARG},\s++({ARG})/index($2, "evict") != -1 && index($2, "queue") == -1 ? "$1($2)" : undef' -tag "WT_WITH_PASS_LOCK(evict_pass_lock)=WT_WITH_PASS_LOCK" -x /_curhs_cache -x /_verbose_dump_cache -x /_cache_destroy -x /_cache_create -d 11 -dir td -dot | dot -Tsvg /dev/stdin > src/docs/images/locks-evict-example-simple.svg -->
```
[the remark this answers is not in the document]
Comment (ershov) [reply src/docs/arch-locking-hierarchy.dox:89]: This is a good point.

I haven't been making any concrete plans to auto-generate these images yet. In the first place, because the call graph tool is not an official one.

Making it official (if we even want it) is a completely different topic outside of this PR. If it ever happens, it's quite easy to switch to auto-generation. For now, it's just a reminder on how to do it manually.

Alternatively, I can move it to an open repo so that it's accessible as a third-party tool so that we can use it but don't have to put it in this repo. Again, we can discuss some approaches to this as another ticket.

[the remark this answers is not in the document]
Comment (ershov) [reply src/docs/arch-locking-hierarchy.dox:89]: BTW here's a command to get all locks descriptions:
```
egrep -Hnr -E '\<[a-zA-Z_0-9]*_lock;' src --include=\*.h
```

File: src/docs/arch-eviction.dox (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/docs/arch-eviction.dox:96]: Okay, that's fine.  No need to do more.

File: src/docs/arch-locking-hierarchy.dox (hunk 8 above)
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/docs/arch-locking-hierarchy.dox:89]: Thanks - I think it's up to you whether you want to make callgraph official.
